File: LeetCode/57.Insert_Interval/57.Insert_Interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new ArrayList<>();
        int l = newInterval[0], r = newInterval[1];
        boolean merge = false;
        for (int[] i : intervals) {
            if (r < i[0]) {
                if (!merge) {
                    res.add(new int[] { l, r });
                    merge = true;
                }
                res.add(i);
            } else if (i[1] < l) {
                res.add(i);
            } else {
                l = Math.min(i[0], l);
                r = Math.max(i[1], r);
            }
        }

        if (!merge) {
            res.add(new int[] { l, r });
        }

        return res.toArray(new int[res.size()][]);
    }
}

File: LeetCode/57.Insert_Interval/57.Insert_Interval.py
#!/usr/bin/env python3


class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        # iterate interval in intervals,
        res = []
        inserted = False
        for cur in intervals:

            # cur is the left side of new interval
            if cur[1] < newInterval[0]:
                res.append(cur)
            # right side
            elif cur[0] > newInterval[1]:
                if not inserted:
                    res.append(newInterval)
                    inserted = True
                res.append(cur)
            else:
                newInterval[0] = min(cur[0], newInterval[0])
                newInterval[1] = max(cur[1], newInterval[1])

        if not inserted:
            res.append(newInterval)

        return res
